iguana.corecontroller/src/main/java/org/aksw/iguana/cc/utils/IndexedQueryReader.java

```diff
@@ -0,0 +1,165 @@
+    }
+
+    /**
+     * Creates an object that indexes each content in between two of the given separators (including the beginning and
+     * end of the given file). <br/>
+     * @param filepath path to the file
+     * @param separator the separator for each bundle
+     * @throws IOException
+     */
+    private IndexedQueryReader(String filepath, String separator) throws IOException {
+        this.file = new File(filepath);
+        this.indexFile(separator);
+    }
+
+    /**
+     * Returns the indexed content with the given index.
+     * @param index the index of the searched content
+     * @return the searched content
+     * @throws IOException
+     */
+    public String readQuery(int index) throws IOException {
+        // split long value into two integers
+        long length = this.indices.get(index)[1];
+        int upper = (int)(length >> 32);
+        int lower = (int)length;
+
+        // store content into two byte arrays (for the case that the content is >~2GB)
+        byte[] data1 = new byte[lower];
+        byte[] data2 = new byte[upper];
+        String output;
+        try(RandomAccessFile raf = new RandomAccessFile(this.file, "r")) {
+            raf.seek(this.indices.get(index)[0]);
+            raf.read(data1);
+            output = new String(data1, StandardCharsets.UTF_8);
+            raf.read(data2);
+            output += new String(data2, StandardCharsets.UTF_8);
+        }
+        return output;
+    }
+
+    /**
+     * This method returns a list of strings that contains every indexed content.
+     * @return list of lines
+     * @throws IOException
+     */
+    public List<String> readQueries() throws IOException {
+        ArrayList<String> out = new ArrayList<>();
+        for(int i = 0; i < indices.size(); i++) {
+            out.add(this.readQuery(i));
+        }
+        return out;
+    }
+
+    /**
+     * Returns the number of indexed content.
+     * @return number of indexed objects
+     */
+    public int size() {
+        return this.indices.size();
+    }
+
+    /**
+     * Indexes every content in between two of the given separator. The beginning and the end of the file count as
+     * separators too.
+     * @param separator the custom separator
+     * @throws IOException
+     */
+    private void indexFile(String separator) throws IOException {
+        this.indices = new ArrayList<>();
+        try(FileReader fr = new FileReader(this.file, StandardCharsets.UTF_8);
+            BufferedReader br = new BufferedReader(fr)) {
+            // The method needs to know the length of the line ending used in the file to be able to properly calculate
+            // the starting byte position of a line
+            long lastIndex = 0;
+            long currentIndex = 0;
+            int counter = 0;
+            int character;
+            boolean isWhiteSpace = true;
+            while((character = br.read()) != -1) {
```

Comment:
This doesn't take care of seperators like `###$`. It would miss the following: `a####$b`. (Result should actually be "a#" and "b" or (0,2), (6,1) respectively)

I would suggest something like this: 
```java
final char[] sepChars = separator.toCharArray();
            long lastQueryIndex = 0;
            long currentIndex = 0;
            int sepCounter = 0;
            int character;

            while((character = br.read()) != -1) {
                // this might become a sep counter
                boolean maybeSep = (character == (int) sepChars[sepCounter]);
                // this is a completed sep counter
                boolean isSep = maybeSep && (sepCounter + 1 == sepChars.length);

                // Before the leading characters of a separator were read but the current character does not match
                // Check if a shorter part of the tail read contains the leading characters of the separator
                if (!maybeSep && sepCounter > 0) {
                    // offset to where we started reading the separator originally
                    offset_loop:
                    for (int offset = 1; offset <= sepCounter; offset++) {
                        boolean sepSubstrMatch = true;
                        // check if separator with an offset from start is the same as from start
                        for (int i = 0; i <= sepCounter - offset; i++) {
                            if (sepChars[i] != sepChars[i + offset])
                                continue offset_loop;
                        }
                        // found a match, adjust sepCounter and maybeSep
                        sepCounter -= offset;
                        maybeSep = true;
                    }
                }
                
                if (isSep) {
                    this.indices.add(new Long[]{lastQueryIndex, currentIndex - lastQueryIndex - sepChars.length});
                    lastQueryIndex = currentIndex;
                    sepCounter = 0;
                } else if (maybeSep){
                    sepCounter++;
                }
                ++currentIndex;
            }
```

PLease doubl check and test before you use the code.